crypto/zk/artifacts/artifacts.go

```diff
@@ -0,0 +1,175 @@
+Each file is hashed with sha256, which output is publicly known. The hash
+values of each file are hardcoded in the VochainGenesis.
+
+*/
+
+const (
+	// FilenameWitness defines the name of the file of the WitnessCalculator
+	FilenameWitness string = "witness.wasm"
+	// FilenameZkey defines the name of the file of the circom ZKey
+	FilenameZkey string = "circuit.zkey"
+	// FilenameVk defines the name of the VerificationKey.json
+	FilenameVk string = "verificationKey.json"
+)
+
+// CircuitsConfig defines the configuration of the files to be downloaded
+type CircuitsConfig struct {
+	// URL defines the URL from where to download the files
+	URL string
+	// CircuitsPath defines the path from where the files are downloaded
+	CircuitsPath string
+	// LocalDir defines in which directory will be the files
+	// downloaded, under that directory it will follow the CircuitsPath
+	// directories structure
+	LocalDir string
+
+	// WitnessHash contains the expected hash for the file filenameWitness
+	WitnessHash []byte // witness.wasm
+	// ZkeyHash contains the expected hash for the file filenameZkey
+	ZkeyHash []byte // circuit.zkey
+	// VkHash contains the expected hash for the file filenameVk
+	VkHash []byte // verificationKey.json
+}
+
+// DownloadCircuitFiles will download the circuits in the specified path,
+// checking the expected sha256 hash of each file. If the files already exist
+// in the path, it will not download them again but it will check the hash of
+// the existing files.
+func DownloadCircuitFiles(c CircuitsConfig) error {
+	// check if files already exist, if files already exist, check hashes
+	err := checkHashes(c)
+	if err == nil {
+		// files already exist, and match the expected hashes
+		return nil
+	}
+	if err != nil && !strings.Contains(err.Error(), "no such file or directory") {
+		return err
+	}
+
+	// at this point, or files do not exist, or exist with a not expected
+	// hash
+	return downloadFiles(c)
+}
+
+func checkHashes(c CircuitsConfig) error {
+	err := checkHash(c.LocalDir+c.CircuitsPath+FilenameWitness, c.WitnessHash)
+	if err != nil {
+		return err
+	}
+
+	err = checkHash(c.LocalDir+c.CircuitsPath+FilenameZkey, c.ZkeyHash)
+	if err != nil {
+		return err
+	}
+
+	err = checkHash(c.LocalDir+c.CircuitsPath+FilenameVk, c.VkHash)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkHash(path string, expected []byte) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+	hash := h.Sum(nil)
+
+	if !bytes.Equal(hash[:], expected) {
+		return fmt.Errorf("path: %s, expected hash: %s, computed hash: %s",
+			path, hex.EncodeToString(expected), hex.EncodeToString(hash[:]))
+	}
+	return nil
+}
+
+// downloadFiles downloads the files, and once downloaded calls checkHashes
+func downloadFiles(c CircuitsConfig) error {
+	err := os.MkdirAll(c.LocalDir+c.CircuitsPath, os.ModePerm)
+	if err != nil {
```

Comment:
why not like this? (same for the following ones):
```go
if err := foo(); err ! nil {
```

Reply:
I didn't put it in this way in purpose because if not the lines get too long and need to be cropped at some point and gets even more confusing, for example:
```go
if err := downloadFile(c.URL+c.CircuitsPath+FilenameWitness, c.LocalDir+c.CircuitsPath+FilenameWitness); err!=nil {
```
this is too long (116 characters), then could be cropped like:
```go
if err := downloadFile(c.URL+c.CircuitsPath+FilenameWitness,
        c.LocalDir+c.CircuitsPath+FilenameWitness); err!=nil {
````
but then is like it takes too long and it's harder to follow.

Then, there are cases where is quite direct to use `if err:= foo(); err!=nil {` as the line is quite short, but then which criteria we use? Like, in the short lines we use it and not in the long ones? or we use it on all the lines even if the code gets more obfuscated and harder to read?

Reply:
I personally prefer always using `if err:= foo(); err!=nil {` even if the line becomes long and I have to split it.  But I totally understand your point.  So for me this conversation is resolved in the sense that we don't have a specific criteria.  I feel this is a very personal opinion.  Maybe @mvdan can give an opinion here?

Reply:
I agree with @ed255. If the line is too long, the added `if ...; err != nil {` doesn't really make it significantly worse. And you should split up the func call between multiple lines, anyway.

You can also do it in a way that's readable:

```
if err := downloadFile(ctx,
        c.URL+c.CircuitsPath+FilenameWitness,
        c.LocalDir+c.CircuitsPath+FilenameWitness,
); err != nil {
```